lints/mozilla/lint_mp_ecdsa_allowed_curve_hash_pair.go

```diff
@@ -0,0 +1,131 @@
+ */
+
+/********************************************************************
+Section 5.1 - Algorithms
+ECDSA keys using one of the following curve-hash pairs: P‐256 with SHA-256, P‐384 with SHA-384
+********************************************************************/
+
+package lints
+
+import (
+	"encoding/asn1"
+	"errors"
+	"fmt"
+	"math/big"
+
+	"github.com/zmap/zcrypto/x509"
+	"github.com/zmap/zlint/lint"
+	"github.com/zmap/zlint/util"
+)
+
+type ecdsaSignature struct {
+	R, S *big.Int
+}
+
+func getSigningKeySize(cert *x509.Certificate) (int, error) {
+	sig := new(ecdsaSignature)
+	if _, err := asn1.Unmarshal(cert.Signature, sig); err != nil {
+		return -1, err
+	}
+
+	rsize := sig.R.BitLen()
+	ssize := sig.S.BitLen()
+
+	switch {
+	case rsize <= 112 && ssize <= 112:
+		return 112, nil
+	case rsize <= 128 && ssize <= 128:
+		return 128, nil
+	case rsize <= 160 && ssize <= 160:
+		return 160, nil
+	case rsize <= 192 && ssize <= 192:
+		return 192, nil
+	case rsize <= 224 && ssize <= 224:
+		return 224, nil
+	case rsize <= 239 && ssize <= 239:
+		return 239, nil
+	case rsize <= 256 && ssize <= 256:
+		return 256, nil
+	case rsize <= 320 && ssize <= 320:
+		return 320, nil
+	case rsize <= 384 && ssize <= 384:
+		return 384, nil
+	case rsize <= 512 && ssize <= 512:
+		return 512, nil
+	case rsize <= 521 && ssize <= 521:
+		return 521, nil
+	}
+
+	return -1, errors.New("cannot identify signing ECDSA key length")
+}
+
+type ecdsaAllowedAlgorithm struct{}
+
+func (l *ecdsaAllowedAlgorithm) Initialize() error {
+	return nil
+}
+
+var applicableSigAlgs = map[x509.SignatureAlgorithm]bool{
+	x509.ECDSAWithSHA1:   true,
+	x509.ECDSAWithSHA256: true,
+	x509.ECDSAWithSHA384: true,
+	x509.ECDSAWithSHA512: true,
+}
+
+func (l *ecdsaAllowedAlgorithm) CheckApplies(c *x509.Certificate) bool {
+	return applicableSigAlgs[c.SignatureAlgorithm]
+}
+
+func (l *ecdsaAllowedAlgorithm) Execute(c *x509.Certificate) *lint.LintResult {
+	signKeySize, err := getSigningKeySize(c)
+	if err != nil {
+		return &lint.LintResult{
+			Status:  lint.Fatal,
+			Details: "cannot identify signing ECDSA key length",
+		}
+	}
+
+	switch {
+	case c.SignatureAlgorithm == x509.ECDSAWithSHA256 && signKeySize == 256:
+		return &lint.LintResult{
+			Status:  lint.Pass,
+			Details: "ECDSAWithSHA256 and 256 bit signing key.",
+		}
+	case c.SignatureAlgorithm == x509.ECDSAWithSHA384 && signKeySize == 384:
+		return &lint.LintResult{
+			Status:  lint.Pass,
+			Details: "ECDSAWithSHA384 and 384 bit signing key.",
+		}
+	}
+
+	return &lint.LintResult{
+		Status: lint.Error,
+		Details: fmt.Sprintf(
+			"Signing key size (%d) did not match sig. alg curve/hash pair specified (%d)",
+			signKeySize,
+			c.SignatureAlgorithm),
+	}
+}
+
+func init() {
+	lint.RegisterLint(&lint.Lint{
+		Name:          "e_mp_ecdsa_allowed_curve_hash_pair",
+		Description:   "ECDSA keys using one of the following curve/hash pairs: P‐256 with SHA-256, P‐384 with SHA-384",
```

Comment:
This description doesn't make much sense. Keys doesn't use a curve/hash pair. Should probably be something about signatures can't use mismatched hash and key type.

Reply:
Something scares me about the logic here, but I'm not sure if that's legitimate. Tagging @dadrian to take a look as well.

Reply:
Agreed. This is coming back to the lengthy discussion sparked from your comment [here](https://github.com/zmap/zlint/pull/323#discussion_r333788201) on @fotisl's original PR. Nobody has been 100% confident in the name/description/approach used. My update to the desc is pretty clumsy :-/

I think an approach similar to what @tadukurow suggested in that thread makes more sense but is a larger diff. Would you prefer we drop this lint entirely and try to do something like that in a follow-up? I'm hesitant to add new code to this PR and would prefer we try to merge as much of it as we feel good about and start fresh with smaller follow-ups.

Reply:
I'm leaning towards excluding it for now, but that's not a super strongly held view.

Reply:
WFM. I'm going to do that to keep momentum.

My plan (not yet articulated anywhere) is to land as much of this PR as possible while filing issues for the things we're removing. That will free me up to switch my attention to #344 and cutting a release. Afterwards I'd like to do a full pass through the Mozilla reqs and file issues for the places we're missing lint coverage above and beyond what was flagged from this PR.

Reply:
Removed in 69fdd4e